pkg/resourceapply/apps_test.go

```diff
@@ -0,0 +1,376 @@
+package resourceapply
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes/fake"
+	appsv1listers "k8s.io/client-go/listers/apps/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/utils/pointer"
+)
+
+func TestApplyStatefulSet(t *testing.T) {
+	// Using a generating function prevents unwanted mutations.
+	newSts := func() *appsv1.StatefulSet {
+		return &appsv1.StatefulSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "default",
+				Name:      "test",
+				// Setting a RV make sure it's propagated to update calls for optimistic concurrency.
+				ResourceVersion: "42",
+				Labels:          map[string]string{},
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						Controller:         pointer.BoolPtr(true),
+						UID:                "abcdefgh",
+						APIVersion:         "scylla.scylladb.com/v1",
+						Kind:               "ScyllaCluster",
+						Name:               "basic",
+						BlockOwnerDeletion: pointer.BoolPtr(true),
+					},
+				},
+			},
+			Spec: appsv1.StatefulSetSpec{
+				Replicas: pointer.Int32Ptr(3),
+				Template: corev1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{},
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{
+							{
+								Name:  "scylla",
+								Image: "scylladb/scylla:latest",
+							},
+						},
+					},
+				},
+			},
+		}
+	}
+
+	newStsWithHash := func() *appsv1.StatefulSet {
+		sts := newSts()
+		utilruntime.Must(SetHashAnnotation(sts))
+		return sts
+	}
+
+	tt := []struct {
+		name            string
+		existing        []runtime.Object
+		cache           []runtime.Object // nil cache means autofill from the client
+		required        *appsv1.StatefulSet
+		expectedSts     *appsv1.StatefulSet
+		expectedChanged bool
+		expectedErr     error
+		expectedEvents  []string
+	}{
+		{
+			name:            "creates a new sts when there is none",
+			existing:        nil,
+			required:        newSts(),
+			expectedSts:     newStsWithHash(),
+			expectedChanged: true,
+			expectedErr:     nil,
+			expectedEvents:  []string{"Normal StatefulSetCreated StatefulSet default/test created"},
```

Comment:
nit: these strings could be a const

Reply:
Usually in tests that verify some functionality I prefer inlining the expected values for safety and if changed, to make it clear what all changes. If you have a constant or variable it may be harder to realize the impact for the author and the reviewer.